src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs

```diff
@@ -0,0 +1,136 @@
+{
+    /// <summary>
+    /// An <see cref="IClientMessageInspector"/> implementation which adds telemetry to outgoing requests.
+    /// </summary>
+    public class TelemetryClientMessageInspector : IClientMessageInspector
+    {
+        /// <inheritdoc/>
+        public object BeforeSendRequest(ref Message request, IClientChannel channel)
+        {
+            try
+            {
+                if (WcfInstrumentationActivitySource.Options == null || WcfInstrumentationActivitySource.Options.OutgoingRequestFilter?.Invoke(request) == false)
+                {
+                    WcfInstrumentationEventSource.Log.RequestIsFilteredOut(request.Headers.Action);
+                    return new State
+                    {
+                        SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                    };
+                }
+            }
+            catch (Exception ex)
+            {
+                WcfInstrumentationEventSource.Log.RequestFilterException(ex);
+                return new State
+                {
+                    SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                };
+            }
+
+            Activity activity = WcfInstrumentationActivitySource.ActivitySource.StartActivity(
+                WcfInstrumentationActivitySource.OutgoingRequestActivityName,
+                ActivityKind.Client);
+
+            IDisposable suppressionScope = this.SuppressDownstreamInstrumentation();
+
+            if (activity != null)
+            {
+                string action = request.Headers.Action;
+                activity.DisplayName = action;
+
+                WcfInstrumentationActivitySource.Options.Propagator.Inject(
+                    new PropagationContext(activity.Context, Baggage.Current),
+                    request,
+                    WcfInstrumentationActivitySource.MessageHeaderValueSetter);
+
+                if (activity.IsAllDataRequested)
+                {
+                    int lastIndex = action.LastIndexOf('/');
+
+                    activity.SetTag(WcfInstrumentationConstants.RpcSystemTag, WcfInstrumentationConstants.WcfSystemValue);
+                    activity.SetTag(WcfInstrumentationConstants.RpcServiceTag, action.Substring(0, lastIndex));
+                    activity.SetTag(WcfInstrumentationConstants.RpcMethodTag, action.Substring(lastIndex + 1));
+                    activity.SetTag(WcfInstrumentationConstants.NetPeerNameTag, channel.RemoteAddress.Uri.Host);
+                    activity.SetTag(WcfInstrumentationConstants.NetPeerPortTag, channel.RemoteAddress.Uri.Port);
+
+                    if (WcfInstrumentationActivitySource.Options.SetSoapVersion)
+                    {
+                        activity.SetTag(WcfInstrumentationConstants.SoapVersionTag, request.Version.ToString());
+                    }
+
+                    activity.SetTag(WcfInstrumentationConstants.WcfChannelSchemeTag, channel.RemoteAddress.Uri.Scheme);
```

Comment:
Either the Via or To header from the message might be better to use here here. If they are using manual addressing, channel.RemoteAddress might not be where the message is headed. Via is the remote endpoint the request is physically being sent to, To is the ultimate endpoint the address is to be delivered to. They usually match but don't always, and don't have to.

Reply:
Good tip thanks. Updated.